### PR TITLE
Spark: Spark SQL Extensions for branch

### PR DESCRIPTION
**Pull request by hililiwei**
Co-authored-by: liliwei [liliwei14@huawei.com](mailto:liliwei14@huawei.com)
Co-authored-by: xuwei [xuwei132@huawei.com](mailto:xuwei132@huawei.com)
Co-authored-by: chidayong [chidayong2@h-partners.com](mailto:chidayong2@h-partners.com)

base on https://github.com/apache/iceberg/pull/5209#issuecomment-1184719859,  Split https://github.com/apache/iceberg/pull/5209  into TAG(#5281) and Branch(#5535)

## What is the purpose of the change
Implement the syntax in the following documents:
https://docs.google.com/document/d/1tbATFPrKF3vNlzkgZQdaW8CAJmbjvryfrlg6C2Ci_aA/edit

#### Create  Branches
```sql
ALTER TABLE tableName
{CREATE | REPLACE} BRANCH branchName [AS OF {VERSION snapshotId}]
[WITH SNAPSHOT RETENTION {[num_snapshots SNAPSHOTS]
[interval {MONTHS | DAYS | HOURS | MINUTES}]}]}]
[RETAIN interval {MONTHS | DAYS | HOURS | MINUTES}]

ALTER TABLE prod.db.my_table 
CREATE BRANCH test_branch AS OF VERSION 100 WITH SNAPSHOT RETENTION 5
SNAPSHOTS 7 DAYS
RETAIN 12 MONTHS

```

#### Drop Branches

```sql
ALTER TABLE table DROP BRANCH reference
```
#### Updating Branch Snapshot Retention

```sql
ALTER TABLE table 
ALTER BRANCH branchName
SET SNAPSHOT RETENTION
{[numSnapshots SNAPSHOTS] [interval {MONTHS | DAYS | HOURS | MINUTES}]}


ALTER TABLE prod.db.my_table 
ALTER BRANCH branchName
SET SNAPSHOT RETENTION
5 SNAPSHOTS 10 DAYS
```

#### Update Retention for Branches
```sql
ALTER TABLE table 
ALTER BRANCH refName
RETAIN interval {MONTHS | DAYS | HOURS | MINUTES}
```

#### Rename Branches
```sql
ALTER TABLE table RENAME BRANCH oldBranch TO newBranch
```

### CODE REVIEW

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 1]
```diff
@@ -73,6 +73,12 @@ statement
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
     | ALTER TABLE multipartIdentifier SET IDENTIFIER_KW FIELDS fieldList                    #setIdentifierFields
     | ALTER TABLE multipartIdentifier DROP IDENTIFIER_KW FIELDS fieldList                   #dropIdentifierFields
+    | ALTER TABLE multipartIdentifier CREATE BRANCH identifier (AS OF VERSION snapshotId)?  (WITH SNAPSHOT RETENTION (((numSnapshots SNAPSHOTS) | (snapshotRetain snapshotRetainTimeUnit)) | (numSnapshots SNAPSHOTS snapshotRetain snapshotRetainTimeUnit)))? (RETAIN snapshotRefRetain snapshotRefRetainTimeUnit)?   #createBranch
+    | ALTER TABLE multipartIdentifier REPLACE BRANCH identifier (AS OF VERSION snapshotId)?  (WITH SNAPSHOT RETENTION (((numSnapshots SNAPSHOTS) | (snapshotRetain snapshotRetainTimeUnit)) | (numSnapshots SNAPSHOTS snapshotRetain snapshotRetainTimeUnit)))? (RETAIN snapshotRefRetain snapshotRefRetainTimeUnit)?   #replaceBranch
```
Comment (kbendick) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:77]: For `with snapshot retention` statement, would it be possible to use something similar to Spark's `temporalClause` in its sql base parser? Maybe `snapshotRetentionClause`?

Having a logical grouping for these would help readability if possible.

Copied from Spark's SqlBaseParser.g4

```
temporalClause
    : FOR? (SYSTEM_VERSION | VERSION) AS OF version=(INTEGER_VALUE | STRING)
    | FOR? (SYSTEM_TIME | TIMESTAMP) AS OF timestamp=valueExpression
    ;
```

Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: good. thx.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala [hunk 2]
```diff
@@ -81,6 +87,85 @@ class IcebergSqlExtensionsAstBuilder(delegate: ParserInterface) extends IcebergS
...
+  override def visitCreateBranch(ctx: CreateBranchContext): CreateBranch = withOrigin(ctx) {
+    val snapshotRetention = Option(ctx.snapshotRetentionClause())
+
+    CreateBranch(
+      typedVisit[Seq[String]](ctx.multipartIdentifier),
+      ctx.identifier().getText,
+      Option(ctx.snapshotId()).map(_.getText.toLong),
+      snapshotRetention.flatMap(s => Option(s.numSnapshots())).map(_.getText.toLong),
+      snapshotRetention.flatMap(s => Option(s.snapshotRetain())).map(
+        _.getText.toLong * timeUnit(ctx.snapshotRetentionClause().snapshotRetainTimeUnit().getText)),
```
Comment (kbendick) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:102]: What will happen here if the `snapshotRetentionTimeUnit()` doesn’t resolve?

Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:102]: The `snapshotRetentionTimeUnit` and `snapshotRetain` are used together. If `snapshotRetain` is unavailable, the `flatMap` will return a `None`, so the `map` does not execute.

**Comment by hililiwei** [comment]
Could you please take a look when you are avaliable? cc @amogh-jahagirdar @jackye1995 @rdblue

**Pull request closed without merging**